File: ocaml/bin/p0070.ml
(* Project Euler: Problem 70 *)

(*
  The answer must be a composite number because prime 'n' is not a permutation of phi(n) = n - 1.

  n = p1^k1 * p2^k2 * p3^k3 * ... * p{r}^k{n}
  -->
    phi(N) = N * (1-1/p1) * (1-1/p2) * (1-1/p3) * ... * (1-1/p{n})
      <-->
    N/phi(N) = (p1/(p1-1)) * (p2/(p2-1)) * (p3/(p3-1)) * ... * (p{n}/(p{n}-1))

  From the problem statement, 87109/phi(87109) = 87109 / 79180 = 1.1001.
  11/10 = 1.1 and 7/6 = 1.666..., so 11 <= prime numbers <= 9_999_999 / 11 = 909090.8181...

  The answer N has the following form (p_i are prime numbers)

    N = p1^k1 * p2^k2 * ... * pn^kn  (N < 10^7, n > 1, 11 <= p1 < p2 < ... < pn, k1>2 when n=1)
 *)

open Core

module E = Euler.Eratosthenes

let u_limit = 9_999_999

let prod pf_lst =
  List.fold pf_lst ~init:1 ~f:(fun acc (base, exp) -> acc * (Int.pow base exp))

let phi pf_lst =
  List.fold pf_lst ~init:1
    ~f:(fun acc (base, exp) -> acc * (Int.pow base (pred exp)) * (pred base))

let get_ratio pf_lst =
  Float.(of_int (prod pf_lst) /. of_int (phi pf_lst))

let pf_generator prime_t tpl =
  (*
    Note:
      The internal data 'pf_lst' has the following structure.
        [(p_n, e_n); ...; (p2, e2); (p1, e1)]
      In contrast, the function 'next' returns its reversed list.
        [(p1, e1); (p2, e2); ...; (p_n, e_n)]
   *)
  let pf_lst = ref (if fst tpl = snd tpl then
                      [(fst tpl, 2)]
                    else
                      [(fst tpl, 1); (snd tpl, 1)]) in
  let aux pf_lst =
    let base_n, exp_n = List.hd_exn pf_lst in
    let tmp = u_limit / (prod pf_lst) in
    if tmp < base_n then
      pf_lst
    else (
      let p_prime = E.prev_prime prime_t (tmp + 1) in
      if p_prime > base_n then
        (p_prime, 1) :: pf_lst
      else
        (base_n, (succ exp_n)) :: (List.tl_exn pf_lst)
    )
  in
  let next () =
    let base_n, exp_n = List.hd_exn !pf_lst in
    if List.length !pf_lst = 1 && exp_n = 1 then
      (* [(p_1, 1)] ->  go to next smaller p_1 *)
      None
    else (
      let result = List.rev !pf_lst in
      if exp_n > 1 then (
        (* [(p_n, e_n); ...] --> [(p_n, e_n - 1); ...] *)
        pf_lst := (base_n, (pred exp_n)) :: (List.tl_exn !pf_lst);
      ) else (
        let p_prime = E.prev_prime prime_t base_n and
            base_n', exp_n' = List.nth_exn !pf_lst 1 in
        if p_prime = base_n' then
          (* [(p_n, 1); (p_{n-1}, e_{n-1}); ...]
               --> [(p_{n-1}, e_{n-1} + 1); ...] *)
          pf_lst := aux ((base_n', (succ exp_n')) :: (List.tl_exn (List.tl_exn !pf_lst)))
        else
          (* [(p_n, 1); (p_{n-1}, e_{n-1}); ...]
               --> [(prev_prime(p_{n}), 1); (p_{n-1}, e_{n-1}; ...] *)
          pf_lst := aux ((p_prime, 1) :: (List.tl_exn !pf_lst))
      );
      Some result
    )
  in
  next

let solve prime_t tpl_lst =
  let module PQ = Euler.PrioQueue.Make(struct
                      type t = float * (int * int) list
                      let compare x y = Float.compare (fst y) (fst x)
                    end) in
  let pq = PQ.init () in

  (* initial data for pruning: phi(87109) = 79180, 87109 = 11 * 7919 *)
  PQ.insert pq (87109. /. 79180., [(11, 1); (7919, 1)]);

  let rec aux pf_gen =
    match pf_gen () with
      None -> true    (* go to next smaller p_1 *)
    | Some pf_lst ->
       if Float.(fst (PQ.peek pq) < get_ratio (List.take pf_lst 2)) then
         true    (* pruning: skip to next smaller p_1 *)
       else (
         if Euler.Math.is_permutation (prod pf_lst) (phi pf_lst) then (
           PQ.insert pq (get_ratio pf_lst, pf_lst)
         );
         aux pf_gen
       )
  in
  ignore (
    List.for_all tpl_lst
      ~f:(fun tpl ->
        let ratio = Float.((of_int (snd tpl) /. of_int (Int.((snd tpl) - 1)))) in
        if Float.(ratio > fst (PQ.peek pq)) then
          false  (* pruning: the end of search *)
        else
          aux (pf_generator prime_t tpl))
  );
  PQ.peek pq

let exec () =
  let prime_t = E.generate ((u_limit / 11) + 1) in
  let init_lst = E.to_list (E.generate (Euler.Math.isqrt u_limit))
                 |> List.drop_while ~f:(fun n -> n < 11)
                 |> List.rev_filter_map
                      ~f:(fun p1 -> let p2 = E.prev_prime prime_t ((u_limit / p1) + 1) in
                                    if p2 >= p1 then Some (p2, p1) else None)
  in
  let ratio, p_lst = solve prime_t init_lst in
  sprintf "%d (n/phi(n): %.10f)" (prod p_lst) ratio

let () = Euler.Task.run exec

File: rust/src/bin/p0070.rs
// Project Euler: Problem 70

/*
  The answer must be a composite number because prime 'n' is not a permutation of phi(n) = n - 1.

  n = p1^k1 * p2^k2 * p3^k3 * ... * p{r}^k{n}
  -->
    phi(N) = N * (1-1/p1) * (1-1/p2) * (1-1/p3) * ... * (1-1/p{n})
      <-->
    N/phi(N) = (p1/(p1-1)) * (p2/(p2-1)) * (p3/(p3-1)) * ... * (p{n}/(p{n}-1))

  From the problem statement, 87109/phi(87109) = 87109 / 79180 = 1.1001.
  11/10 = 1.1 and 7/6 = 1.666..., so 11 <= prime numbers <= 9_999_999 / 11 = 909090.8181...

  The answer N has the following form (p_i are prime numbers)

    N = p1^k1 * p2^k2 * ... * pn^kn  (N < 10^7, n > 1, 11 <= p1 < p2 < ... < pn, k1>2 when n=1)
*/

use euler::math::{self, primes};
use std::{cmp::Ordering, collections::BinaryHeap};

euler::run_solver!(70);

fn solve() -> String {
    compute().to_string()
}

fn compute() -> i64 {
    use std::cmp;

    let limit = 10_000_000 - 1;
    let mut pq: BinaryHeap<Ratio> = BinaryHeap::new();
    pq.push(Ratio {
        priority: ratio(87109, 79180),
        num: 87109,
    });

    for p in primes::primes(11, math::isqrt(limit)).into_iter().rev() {
        if get_phi_ratio(&[(p, 1)]) > pq.peek().unwrap().priority {
            break;
        }

        for pf_lst in PrimeFactorization::new((p, primes::prev_prime(limit / p + 1)), limit) {
            if get_phi_ratio(&pf_lst[0..cmp::min(2, pf_lst.len())]) > pq.peek().unwrap().priority {
                break;
            }
            if is_perm(prod(&pf_lst), phi(&pf_lst)) {
                pq.push(Ratio {
                    priority: get_phi_ratio(&pf_lst),
                    num: prod(&pf_lst),
                });
            }
        }
    }

    pq.peek().unwrap().num
}

fn prod(pf_lst: &[(i64, i64)]) -> i64 {
    pf_lst.iter().fold(1, |acc, x| acc * (x.0).pow(x.1 as u32))
}

fn phi(pf_lst: &[(i64, i64)]) -> i64 {
    pf_lst
        .iter()
        .fold(1, |acc, x| acc * (x.0).pow(x.1 as u32 - 1) * (x.0 - 1))
}

fn get_phi_ratio(pf_lst: &[(i64, i64)]) -> f32 {
    ratio(prod(pf_lst), phi(pf_lst))
}

fn ratio(num: i64, denom: i64) -> f32 {
    num as f32 / denom as f32
}

// minimum binary heap (priority key: n/phi(n), value: n)
#[derive(Debug, Clone, PartialEq)]
struct Ratio {
    priority: f32,
    num: i64,
}

impl Eq for Ratio {}

impl Ord for Ratio {
    fn cmp(&self, other: &Self) -> Ordering {
        other.priority.partial_cmp(&self.priority).unwrap()
    }
}

impl PartialOrd for Ratio {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

// prime factorization of candidates
#[derive(Debug)]
struct PrimeFactorization {
    pf_lst: Vec<(i64, i64)>,
    limit: i64,
}

impl PrimeFactorization {
    fn new(tpl: (i64, i64), limit: i64) -> PrimeFactorization {
        if tpl.0 == tpl.1 {
            PrimeFactorization {
                pf_lst: vec![(tpl.0, 2)],
                limit,
            }
        } else {
            PrimeFactorization {
                pf_lst: vec![(tpl.1, 1), (tpl.0, 1)],
                limit,
            }
        }
    }
}

impl Iterator for PrimeFactorization {
    type Item = Vec<(i64, i64)>;

    fn next(&mut self) -> Option<Vec<(i64, i64)>> {
        fn aux(lst: &[(i64, i64)], limit: i64) -> Vec<(i64, i64)> {
            let (b, e) = lst[0];
            let tmp = limit / prod(lst);
            if tmp < b {
                lst.to_vec()
            } else {
                let prev_p = primes::prev_prime(tmp);
                if prev_p > b {
                    [&[(prev_p, 1_i64)], lst].concat()
                } else {
                    [&[(b, e + 1)], &lst[1..]].concat()
                }
            }
        }

        let (b, e) = self.pf_lst[0];
        if self.pf_lst.len() == 1 && e == 1 {
            return None;
        }

        let mut result = self.pf_lst.clone();
        result.reverse();
        if e > 1 {
            self.pf_lst[0] = (b, e - 1);
        } else {
            let (new_b, new_e) = self.pf_lst[1];
            let prev_p = primes::prev_prime(b);
            if prev_p == new_b {
                self.pf_lst = aux(
                    &[&[(new_b, new_e + 1)], &self.pf_lst[2..]].concat(),
                    self.limit,
                );
            } else {
                self.pf_lst = aux(
                    &[&[(prev_p, 1_i64)], &self.pf_lst[1..]].concat(),
                    self.limit,
                );
            }
        }

        Some(result)
    }
}

fn is_perm(x: i64, y: i64) -> bool {
    let mut tmp_x = math::digits(x);
    let mut tmp_y = math::digits(y);

    tmp_x.sort_unstable();
    tmp_y.sort_unstable();
    tmp_x == tmp_y
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0070() {
        assert_eq!(compute(), 8319823);
    }
}
